include/pybind11/detail/eigen_tensor.h

```diff
@@ -0,0 +1,401 @@
+        static constexpr descr<sizeof...(Is) * 3 - 2> value
+            = concat(const_name(((void) Is, "?"))...);
+    };
+
+    static constexpr auto dimensions_descriptor
+        = helper<decltype(make_index_sequence<T::NumIndices>())>::value;
+
+    template <typename... Args>
+    static T *alloc(Args &&...args) {
+        return new T(std::forward<Args>(args)...);
+    }
+
+    static void free(T *tensor) { delete tensor; }
+};
+
+template <typename Scalar_, typename std::ptrdiff_t... Indices, int Options_, typename IndexType>
+struct eigen_tensor_helper<
+    Eigen::TensorFixedSize<Scalar_, Eigen::Sizes<Indices...>, Options_, IndexType>> {
+    using T = Eigen::TensorFixedSize<Scalar_, Eigen::Sizes<Indices...>, Options_, IndexType>;
+    using ValidType = void;
+
+    static constexpr Eigen::DSizes<typename T::Index, T::NumIndices> get_shape(const T & /*f*/) {
+        return get_shape();
+    }
+
+    static constexpr Eigen::DSizes<typename T::Index, T::NumIndices> get_shape() {
+        return Eigen::DSizes<typename T::Index, T::NumIndices>(Indices...);
+    }
+
+    static bool is_correct_shape(const Eigen::DSizes<typename T::Index, T::NumIndices> &shape) {
+        return get_shape() == shape;
+    }
+
+    static constexpr auto dimensions_descriptor = concat(const_name<Indices>()...);
+
+    template <typename... Args>
+    static T *alloc(Args &&...args) {
+        Eigen::aligned_allocator<T> allocator;
+        return ::new (allocator.allocate(1)) T(std::forward<Args>(args)...);
+    }
+
+    static void free(T *tensor) {
+        Eigen::aligned_allocator<T> allocator;
+        tensor->~T();
+        allocator.deallocate(tensor, 1);
+    }
+};
+
+template <typename T>
+struct get_tensor_descriptor {
+    static constexpr auto value
+        = const_name("numpy.ndarray[") + npy_format_descriptor<typename T::Scalar>::name
+          + const_name("[") + eigen_tensor_helper<T>::dimensions_descriptor
+          + const_name("], flags.writeable, ")
+          + const_name<(int) T::Layout == (int) Eigen::RowMajor>("flags.c_contiguous",
+                                                                 "flags.f_contiguous");
+};
+
+template <typename Type>
+struct type_caster<Type, typename eigen_tensor_helper<Type>::ValidType> {
+    using H = eigen_tensor_helper<Type>;
+    PYBIND11_TYPE_CASTER(Type, get_tensor_descriptor<Type>::value);
+
+    bool load(handle src, bool /*convert*/) {
+        array_t<typename Type::Scalar, compute_array_flag_from_tensor<Type>()> a(
+            reinterpret_borrow<object>(src));
+
+        if (a.ndim() != Type::NumIndices) {
+            return false;
+        }
+
+        Eigen::DSizes<typename Type::Index, Type::NumIndices> shape;
+        std::copy(a.shape(), a.shape() + Type::NumIndices, shape.begin());
+
+        if (!H::is_correct_shape(shape)) {
+            return false;
+        }
+
+        value = Eigen::TensorMap<Type>(const_cast<typename Type::Scalar *>(a.data()), shape);
```

Comment:
Just a note that this will always assume the data pointer is unaligned (which is maybe the best we can guarantee).  This may have performance implications though.

Reply:
It's a compile-time option in the map, e.g.
```
using AlignedTensorMap = Eigen::TensorMap<Type, Eigen::Aligned>;
```
I'm not sure if we can make guarantees about alignment of the incoming src data (I'm an Eigen maintainer, not a pybind11 user), or if you can set some kind of option.

Reply:
FWIW: In the interest of getting this PR merged, I'd be OK to just adding a comment here, something like: `// Potentially this could be optimized assuming alignment.`

Reply:
We can't know at compile time if incoming data is aligned, but 99.99% of the time it will be aligned and we could theoretically check at runtime. Does Eigen already do a runtime check? Would adding one to our code be redundant?

Seems like probably a good idea if Eigen isn't already doing a runtime check.

Reply:
Numpy uses malloc and has no additional alignment machinery, so it will be aligned most of the time as amd64 defines malloc as 16-byte aligned.

> Sounds safer to keep it unaligned then. 

What would be wrong with doing a simple `std::size_t(result)%EIGEN_DEFAULT_ALIGN_BYTES)==0` check?

Reply:
It will fail if the data is allocated by anything other than Eigen.  I assume we don't want this to crash.

Reply:
We would dynamically fall back to the slower, unaligned code path.

Reply:
Sigh, `pre-existing Eigen::Tensor with guaranteed alignment`, I think we can but only in newer version of Eigen with C++17. :/

Reply:
That isn't quite what we are looking for. That gives the alignment of individual elements in the array, not the overall alignment of the entire array.

Reply:
@cantonios I added support for alignment. Can you take a look at https://github.com/pybind/pybind11/pull/4201/commits/9efbda01209a180b66f32b082d11b113e3994713

Reply:
Looks good, I didn't know what where `value` was coming from, but if it's just `Type`, then this should be fine.